src/frontmatter.md

```diff
@@ -0,0 +1,43 @@
+r[frontmatter]
+# Frontmatter
+
+r[frontmatter.syntax]
+```grammar,lexer
+@root FRONTMATTER ->
+      FRONTMATTER_FENCE HORIZONTAL_WHITESPACE* INFOSTRING? HORIZONTAL_WHITESPACE* LF
+      (FRONTMATTER_LINE LF )*
+      FRONTMATTER_FENCE[^matched-fence] HORIZONTAL_WHITESPACE* LF
+
+FRONTMATTER_FENCE -> `---` `-`*
+
+INFOSTRING -> (XID_Start | `_`) ( XID_Continue | `-` | `.` )*
+
+FRONTMATTER_LINE -> (~INVALID_FRONTMATTER_LINE_START (~INVALID_FRONTMATTER_LINE_CONTINUE)*)?
+
+INVALID_FRONTMATTER_LINE_START -> (FRONTMATTER_FENCE[^escaped-fence] | LF)
+
+INVALID_FRONTMATTER_LINE_CONTINUE -> LF
+```
+
+[^matched-fence]: The closing fence must have the same number of `-` as the opening fence
+[^escaped-fence]: A `FRONTMATTER_FENCE` at the beginning of a `FRONTMATTER_LINE` is only invalid if it has the same or more `-` as the `FRONTMATTER_FENCE`
+
+Frontmatter is an optional section for content intended for external tools without requiring these tools to have full knowledge of the Rust grammar.
+
+r[frontmatter.document]
+Frontmatter may only be preceded by a [shebang] and whitespace.
+
+r[frontmatter.fence]
+The delimiters are referred to as a "fence."
+The opening and closing fences must be at the start of a line.
+They must be a matching pair of three or more hyphens (`-`).
+A fence may be followed by horizontal whitespace.
```

Comment:
In the Reference, for new content, we don't hard wrap lines.  Please unwrap these, here and elsewhere.

Reply:
Huh, I thought I had seen a comment about semantic line feeds (which were introduced to me by ehuss).

I'm surprised then by the motivation

> Do not wrap long lines. This helps with reviewing diffs of the source.

For myself, I find it makes reviewing harder.